File: R/Regresión Elastic Net.R

library(readxl)
library(caret)
library(corrplot)
library(goftest)
library(psych)
library(dplyr)
library(glmnet)
library(ggplot2)
library(tidyr)

# Cargar datos
df_series <- read_excel("Datos/df_series.xlsx", sheet = "Datos")
Proyecto_4_Cuentas_de_Captación_2024 <- "Datos/Proyecto 4 Cuentas de Captación 2024.xlsx"

# Función para cargar y filtrar fechas
cargar_filtrar <- function(sheet_name) {
  read_excel(Proyecto_4_Cuentas_de_Captación_2024, sheet = sheet_name) %>%
    mutate(Fecha = as.Date(Fecha)) %>%
    filter(Fecha >= as.Date("2019-12-01") & Fecha <= as.Date("2024-06-01"))
}

# Cargar hojas de cada banco
TBM <- cargar_filtrar("TBM")
BX <- cargar_filtrar("Banamex")
BBVA <- cargar_filtrar("BBVA")
SA <- cargar_filtrar("Santander")
BN <- cargar_filtrar("Banorte")

# Unir con variables macro
TBM2 <- left_join(TBM, df_series, by = "Fecha")
BX2 <- left_join(BX, df_series, by = "Fecha")
BBVA2 <- left_join(BBVA, df_series, by = "Fecha")
SA2 <- left_join(SA, df_series, by = "Fecha")
BN2 <- left_join(BN, df_series, by = "Fecha")

# Crear lista de bancos
bancos <- list(
  Banxico = TBM2,
  Banamex = BX2,
  BBVA = BBVA2,
  Santander = SA2,
  Banorte = BN2
)

# Variables
variables_macro <- colnames(df_series)[-1]
variables_captacion <- c("Captación tradicional", "Depositos Plazo", "Depositos Vista")

# Función del modelo Elastic Net
ajustar_y_analizar_elastic_net <- function(df, banco, variable_respuesta, variables_predictoras, alpha = 0.5) {
  df <- df %>% select(all_of(c("Fecha", variable_respuesta, variables_predictoras))) %>% na.omit()
  x <- scale(as.matrix(df[, variables_predictoras]))
  y <- as.matrix(df[[variable_respuesta]])
  
  if (is.null(y) || var(y) == 0) return(NULL)
  
  # Ajuste con validación cruzada
  cv_lambda <- cv.glmnet(x, y, alpha = alpha)
  lambda_opt <- cv_lambda$lambda.min
  modelo <- glmnet(x, y, alpha = alpha, lambda = lambda_opt)
  y_hat <- predict(modelo, newx = x)
  coef_elastic <- coef(modelo)
  
  # Evaluación
  ss_total <- sum((y - mean(y))^2)
  ss_residual <- sum((y - y_hat)^2)
  r2 <- 1 - (ss_residual / ss_total)
  prueba_spearman <- cor.test(as.numeric(y), as.numeric(y_hat), method = "spearman")
  
  # Resultados en consola
  cat("\n---", banco, "-", variable_respuesta, "---\n")
  cat("R²:", round(r2, 4), "\n")
  cat("Lambda óptimo:", round(lambda_opt, 6), "\n")
  cat("Coeficientes (Elastic Net):\n"); print(coef_elastic)
  cat("Prueba Spearman:\n"); print(prueba_spearman)
  
  cat("Interpretación económica:\n")
  for (v in variables_predictoras) {
    beta <- coef_elastic[v, 1]
    sentido <- ifelse(beta > 0, "positiva", ifelse(beta < 0, "negativa", "nula"))
    cat(v, ": relación", sentido, "\n")
  }
  
  # Dataframe para gráfica
  df_plot <- data.frame(
    Fecha = df$Fecha,
    Valor = c(as.numeric(y), as.numeric(y_hat)),
    Tipo = rep(c("Real", "Estimado"), each = nrow(df)),
    Variable = variable_respuesta,
    stringsAsFactors = FALSE
  )
  
  return(df_plot)
}

# Lista para almacenar resultados por banco
resultados_banco <- list()

for (banco in names(bancos)) {
  df_banco <- bancos[[banco]]
  graficas_var <- list()
  
  for (var_resp in variables_captacion) {
    if (var_resp %in% colnames(df_banco)) {
      df_plot <- ajustar_y_analizar_elastic_net(df_banco, banco, var_resp, variables_macro)
      if (!is.null(df_plot)) {
        graficas_var[[var_resp]] <- df_plot
      }
    }
  }
  
  # Combinar resultados de variables del banco
  resultados_banco[[banco]] <- bind_rows(graficas_var)
}

# Graficar resultados por banco
for (banco in names(resultados_banco)) {
  df_plot <- resultados_banco[[banco]]
  g <- ggplot(df_plot, aes(x = Fecha, y = Valor, color = Variable, linetype = Tipo)) +
    geom_line(size = 1) +
    labs(title = paste("Modelo Elastic Net -", banco, "(3 variables de captación)"),
         y = "Valor", x = "Fecha", color = "Variable", linetype = "Tipo") +
    theme_minimal()
  print(g)
}

File: Python/Regresión Elastic Net.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.linear_model import ElasticNetCV
from sklearn.preprocessing import StandardScaler
from scipy.stats import spearmanr
from pathlib import Path

# ================================
# Cargar datos
# ================================
df_series = pd.read_excel("Datos/df_series.xlsx", sheet_name="Datos")
proyecto_path = Path("Datos/Proyecto 4 Cuentas de Captación 2024.xlsx")

def cargar_filtrar(sheet_name):
    df = pd.read_excel(proyecto_path, sheet_name=sheet_name)
    df["Fecha"] = pd.to_datetime(df["Fecha"])
    return df[(df["Fecha"] >= "2019-12-01") & (df["Fecha"] <= "2024-06-01")]

TBM = cargar_filtrar("TBM")
BX = cargar_filtrar("Banamex")
BBVA = cargar_filtrar("BBVA")
SA = cargar_filtrar("Santander")
BN = cargar_filtrar("Banorte")

# Merge con variables macro
TBM2 = pd.merge(TBM, df_series, on="Fecha", how="left")
BX2 = pd.merge(BX, df_series, on="Fecha", how="left")
BBVA2 = pd.merge(BBVA, df_series, on="Fecha", how="left")
SA2 = pd.merge(SA, df_series, on="Fecha", how="left")
BN2 = pd.merge(BN, df_series, on="Fecha", how="left")

bancos = {
    "Banxico": TBM2,
    "Banamex": BX2,
    "BBVA": BBVA2,
    "Santander": SA2,
    "Banorte": BN2
}

variables_macro = df_series.columns.drop("Fecha").tolist()
variables_captacion = ["Captación tradicional", "Depositos Plazo", "Depositos Vista"]

# ================================
# Función principal
# ================================
def ajustar_y_analizar(df, banco, y_var, x_vars, l1_ratio=0.5):
    df = df[["Fecha", y_var] + x_vars].dropna()
    if df.empty or df[y_var].var() == 0:
        print(f"Datos insuficientes o varianza nula para {banco} - {y_var}")
        return None

    X = df[x_vars].values
    y = df[y_var].values

    # Escalar solo X, no y
    scaler = StandardScaler()
    X_scaled = scaler.fit_transform(X)

    # Definir rango alphas similar a glmnet de R
    alphas_range = np.logspace(-4, 1, 100)

    # Ajuste con ElasticNetCV
    model_cv = ElasticNetCV(l1_ratio=l1_ratio, alphas=alphas_range, cv=5, random_state=42, max_iter=10000)
    model_cv.fit(X_scaled, y)

    y_hat = model_cv.predict(X_scaled)

    ss_total = np.sum((y - np.mean(y)) ** 2)

    ss_residual = np.sum((y - y_hat) ** 2)
    r2 = 1 - ss_residual / ss_totals

    spearman_corr, spearman_p = spearmanr(y, y_hat)

    print(f"\n--- {banco} - {y_var} ---")
    print(f"R²: {r2:.4f}")
    print(f"Lambda óptimo (alpha): {model_cv.alpha_:.6f}")
    print("Coeficientes:")
    for var, coef in zip(x_vars, model_cv.coef_):
        print(f"  {var}: {coef:.6f}")
    print("Prueba Spearman:")
    print(f"  Coeficiente: {spearman_corr:.4f}, p-valor: {spearman_p:.4g}")
    print("Interpretación económica:")
    for var, coef in zip(x_vars, model_cv.coef_):
        sentido = "positiva" if coef > 0 else "negativa" if coef < 0 else "nula"
        print(f"  {var}: relación {sentido}")

    df_plot = pd.DataFrame({
        "Fecha": list(df["Fecha"]) + list(df["Fecha"]),
        "Valor": np.concatenate([y, y_hat]),
        "Tipo": ["Real"] * len(y) + ["Estimado"] * len(y),
        "Variable": [y_var] * len(y) * 2
    })

    return df_plot


# ================================
# Ajuste de modelos por banco
# ================================
sns.set(style="whitegrid")
resultados = {}

for banco, df_banco in bancos.items():
    graficas = []
    for var in variables_captacion:
        if var in df_banco.columns:
            df_plot = ajustar_y_analizar(df_banco, banco, var, variables_macro)
            if df_plot is not None:
                graficas.append(df_plot)
    if graficas:
        resultados[banco] = pd.concat(graficas, ignore_index=True)

# ================================
# Mostrar gráficos por banco
# ================================
for banco, df_plot in resultados.items():
    print(f"\n====== GRÁFICA BANCO: {banco} ======")
    input("Presiona ENTER para mostrar la gráfica...")
    plt.figure(figsize=(10, 6))
    sns.lineplot(data=df_plot, x="Fecha", y="Valor", hue="Variable", style="Tipo", linewidth=2)
    plt.title(f"Modelo Elastic Net - {banco} (3 variables de captación)")
    plt.xlabel("Fecha")
    plt.ylabel("Valor")
    plt.xticks(rotation=45)
    plt.legend(title="Variable / Tipo")
    plt.tight_layout()
    plt.show()

input("\nFin del script. Presiona ENTER para cerrar...")
